ecommerce/extensions/payment/processors/paytabs.py

```diff
@@ -0,0 +1,192 @@
+            return next(
+                (
+                    field.get('field_value', default_val) for field in account_details['extended_profile']
+                    if field['field_name'] == field_name
+                ),
+               default_val
+            )
+
+        account_details = user.account_details(request)
+        return {
+            'first_name': get_extended_field(account_details, 'first_name'),
+            'last_name': get_extended_field(account_details, 'last_name'),
+            'mailing_address': get_extended_field(account_details, 'mailing_address'),
+            'city': get_extended_field(account_details, 'city'),
+            'country_code': account_details.get('country'),
+            'phone_number': get_extended_field(account_details, 'phone_number', ''),
+            'postal_code': get_extended_field(account_details, 'ZIP/Postal Code', '11564'),
+            'state': get_extended_field(account_details, 'state'),
+        }
+
+    def _get_course_id_title(self, line):
+        """
+        Return the line title prefixed with the course ID, if available.
+        """
+        course_id = ''
+        line_course = line.product.course
+        if line_course:
+            course_id = '{}|'.format(line_course.id)
+        return course_id + line.product.title
+
+    def get_transaction_parameters(self, basket, request=None, use_client_side_checkout=False, **kwargs):
+        """
+        Return the transaction parameters needed for this processor.
+        """
+        site_url = '{}://{}'.format(request.scheme, request.get_host())
+
+        # This can only be used in production when the requested site is publicly accessible on the internet
+        # via HTTPS
+        # return_url = request.build_absolute_uri(reverse('paytabs:submit'))
+        return_url = '{}{}'.format(self.return_base_url, reverse('paytabs:submit'))
+        client_ip, is_routable = get_client_ip(request)
+        ip_merchant = '-'
+        ip_customer = client_ip if is_routable else '-'
+        title = 'Order: {}'.format(basket.order_number)
+        user = basket.owner
+        user_profile_data = self._get_user_profile_data(user, request)
+        products = {'names': [], 'quantities': [], 'prices': []}
+        for line in basket.all_lines():
+            products['names'].append(self._get_course_id_title(line))
+            products['quantities'].append(six.text_type(line.quantity))
+            products['prices'].append(six.text_type(line.line_price_incl_tax_incl_discounts / line.quantity))
+
+        invoiced_products = util.prepare_invoice(products)
+        other_charges = '0.0'
+        amount = str(basket.total_incl_tax)
+        discount = '0.0'
+        currency = basket.currency
+        reference_no = basket.order_number
+        msg_lang = request.COOKIES.get(settings.LANGUAGE_COOKIE_NAME) or settings.LANGUAGE_CODE
+        cms_with_version = 'Open edX E-Commerce'
```

Comment:
This string isn't actually used anywhere except PayTabs logs to know what software is using their API. Since this code is specific to ecommerce this won't be subject to change.